### ISSUE TITLE
CloudEvent should enforce lower-case attribute names

### ISSUE THREAD

**Issue by jskeet** [cloudevents/sdk-csharp#59]
The spec states:

> CloudEvents attribute names MUST consist of lower-case letters ('a' to 'z') or digits ('0' to '9') from the ASCII character set. Attribute names SHOULD be descriptive and terse and SHOULD NOT exceed 20 characters in length.

Unfortunately, this can easily be violated:

```csharp
var evt = new CloudEvent(CloudEventsSpecVersion.V1_0, "my-type", new Uri("//my-source"), "my-id", DateTime.UtcNow);
evt.GetAttributes()["Subject"] = "my-subject";

var formatter = new JsonEventFormatter();
var bytes = formatter.EncodeStructuredEvent(evt, out _);
var text = Encoding.UTF8.GetString(bytes);
// Includes "Subject" instead of "subject"
Console.WriteLine(text);
```

I ran into this because the `HttpRequest.ReadCloudEventAsync` extension method in the ASP.NET Core package preserves header casing - so a header of "Ce-Type" ends up with a value of "Type" in the JSON.

I'm ambivalent as to whether we lower-case in the attributes, or reject an attempt to set an invalid attribute name, but we should definitely do one of the two.

**Comment by jskeet** [comment]
Note: spec version 0.1 allowed non-lower-case-only names, but it's been "lower case only" since 0.2.

So this is related to #56. If we can at least drop 0.1, we can go case-sensitive all the way.

**Comment by jskeet** [comment]
Fixes as of #92.
